pubnub-gson/src/main/java/com/pubnub/internal/PubNubImpl.java

```diff
@@ -0,0 +1,591 @@
+    }
+
+    @Override
+    @NotNull
+    public RemoveMessageAction removeMessageAction() {
+        return new RemoveMessageAction(getCorePubNubClient());
+    }
+
+    // End Message Actions API
+
+    @Override
+    public SendFile.Builder sendFile() {
+        return SendFile.builder(getCorePubNubClient());
+    }
+
+    @Override
+    public ListFiles.Builder listFiles() {
+        return new ListFiles.Builder(getCorePubNubClient());
+    }
+
+    @Override
+    public GetFileUrl.Builder getFileUrl() {
+        return GetFileUrl.builder(getCorePubNubClient());
+    }
+
+    @Override
+    public DownloadFile.Builder downloadFile() {
+        return DownloadFile.builder(getCorePubNubClient());
+    }
+
+    @Override
+    public DeleteFile.Builder deleteFile() {
+        return DeleteFile.builder(getCorePubNubClient());
+    }
+
+    @Override
+    public PublishFileMessage.Builder publishFileMessage() {
+        return PublishFileMessage.builder(getCorePubNubClient());
+    }
+
+    @Override
+    public void reconnect() {
+        super.reconnect(0);
+    }
+
+    // public methods
+
+    /**
+     * Perform Cryptographic decryption of an input string using cipher key provided by PNConfiguration
+     *
+     * @param inputString String to be encrypted
+     * @return String containing the encryption of inputString using cipherKey
+     */
+    @Override
+    @Nullable
+    public String decrypt(String inputString) throws PubNubException {
+        return decrypt(inputString, getCorePubNubClient().getConfiguration().getCipherKey());
+    }
+
+    /**
+     * Perform Cryptographic decryption of an input string using the cipher key
+     *
+     * @param inputString String to be encrypted
+     * @param cipherKey   cipher key to be used for encryption
+     * @return String containing the encryption of inputString using cipherKey
+     * @throws PubNubException throws exception in case of failed encryption
+     */
+    @Override
+    @Nullable
+    public String decrypt(String inputString, String cipherKey) throws PubNubException {
+        return getCorePubNubClient().decrypt(inputString, cipherKey);
+    }
+
+    @Override
+    public InputStream decryptInputStream(InputStream inputStream) throws PubNubException {
+        return decryptInputStream(inputStream, getCorePubNubClient().getConfiguration().getCipherKey());
+    }
+
+    @Override
+    public InputStream decryptInputStream(InputStream inputStream, String cipherKey) throws PubNubException {
+        return getCorePubNubClient().decryptInputStream(inputStream, cipherKey);
+    }
+
+    /**
+     * Perform Cryptographic encryption of an input string and the cipher key provided by PNConfiguration
+     *
+     * @param inputString String to be encrypted
+     * @return String containing the encryption of inputString using cipherKey
+     */
+    @Override
+    @Nullable
+    public String encrypt(String inputString) throws PubNubException {
```

Comment:
In current version there is validation like:
```
     if (inputString == null) {
            throw PubNubException.builder().pubnubError(PubNubErrorBuilder.PNERROBJ_INVALID_ARGUMENTS).build();
        }
```
Why not to use it here as well?

Reply:
I'll think about how to approach validation some more, in general we should have it in core as much as possible to not duplicate it in kotlin/java

Reply:
readded for now